src/js_of_ocaml_rules.ml

```diff
@@ -149,20 +75,109 @@ let setup_separate_compilation_rules sctx components =
+            ) else (
+              let lib_name = Lib.name lib in
+              List.map ~f:(fun js ->
+                in_build_dir ~ctx [lib_name ; Path.basename js]) jsoo_archives))
+        in
+        (* Special case for the stdlib because it is not referenced in the META *)
+        let all_libs = in_build_dir ~ctx ["stdlib"; "stdlib.cma.js"] :: all_libs in
+        let all_other_modules =
+          List.map cm ~f:(fun m -> Path.extend_basename m ~suffix:".js")
+        in
+        Arg_spec.Deps (List.concat [all_libs;all_other_modules]))
+    in
+    let jsoo_link = SC.resolve_program sctx ~hint:install_jsoo_hint "jsoo_link" in
+    Build.run ~context:ctx ~dir
+      jsoo_link
+      [ Arg_spec.A "-o"; Target target
+      ; Arg_spec.Dep runtime
+      ; Arg_spec.As (sourcemap sctx)
+      ; Arg_spec.Dyn get_all
+      ]
+
+  let build_cm ~(js_of_ocaml:Jbuild.Js_of_ocaml.t) ~src ~target =
+    if separate_compilation_enabled sctx
+    then
+      let itarget = Path.extend_basename src ~suffix:".js" in
+      let spec = Arg_spec.Dep src in
+      let flags =
+        SC.expand_and_eval_set sctx ~scope ~dir js_of_ocaml.flags
+          ~standard:(Build.return (standard sctx))
+      in
+      [ flags
+        >>>
+        js_of_ocaml_rule sctx ~dir ~flags:(fun flags -> As flags) ~spec ~target:itarget
+      ]
+      @ (if target = itarget then
+           []
+         else
+           [Build.symlink ~src:itarget ~dst:target])
+    else []
+
+
+  let standalone_runtime_rule ~javascript_files ~target =
+    let spec =
+      Arg_spec.S
+        [ Arg_spec.of_result_map requires ~f:(fun libs ->
+            Arg_spec.Deps (Lib.L.jsoo_runtime_files libs))
+        ; Arg_spec.Deps javascript_files
+        ]
+    in
+    Build.arr (fun (cm_files, flags) -> (cm_files, "--runtime-only" :: flags))
+    >>>
+    js_of_ocaml_rule sctx ~dir ~flags:(fun (_,flags) -> As flags) ~target ~spec
+
+  let exe_rule ~javascript_files ~src ~target =
+    let spec =
+      Arg_spec.S
+        [ Arg_spec.of_result_map requires
+            ~f:(fun libs -> Arg_spec.Deps (Lib.L.jsoo_runtime_files libs))
+        ; Arg_spec.Deps javascript_files
+        ; Arg_spec.Dep src
+        ]
+    in
+    js_of_ocaml_rule sctx ~dir ~flags:(fun (_,flags) -> As flags) ~spec ~target
+
+
+  let build_exe ~js_of_ocaml ~src =
+    let {Jbuild.Js_of_ocaml.javascript_files; _} = js_of_ocaml in
+    let javascript_files =
+      List.map javascript_files ~f:(Path.relative dir) in
+    let mk_target ext = Path.extend_basename src ~suffix:ext in
+    let target = mk_target ".js" in
+    let standalone_runtime = mk_target ".runtime.js" in
+    if separate_compilation_enabled sctx then
+      [ link_rule ~runtime:standalone_runtime ~target
+      ; standalone_runtime_rule ~javascript_files ~target:standalone_runtime
+      ]
+    else
+      [ exe_rule ~javascript_files ~src ~target ]
+end
+
+let build_cm cctx =
+  let module Jsoo = Gen(struct let cctx = cctx end) in
```

Comment:
@rgrinberg it's not immediately clear to me what is the cost of doing this. For instance the `Gen` functor will now be applied for every .ml file. It's fine if it only contain functions, but if we start adding toplevel values in the future performances might degrade.

I think we should stick to passing `cctx` explicitly here. Adding `module C = Compilation_context` should be enough to remove most of the clutter.

Reply:
Okay, but I think the menhir also does this. I'll just get rid of the functor, the boilerplate wasn't even that bad.